### ISSUE TITLE
Ingester:  Improve persist throughput  per-Partition

### ISSUE THREAD

**Issue by aierui** [influxdata/influxdb_iox#6146]
# perf issue

During my testing, I found that the Ingester module was very slow and the module did not resource bottlenecks(such as cpu, memory). when the Router had a lot of traffic, the Ingester TTBR metrics would skyrocket and pile up more and more data in the WriteBuffer(redpanda). Based on the inch test tool, some of my key metrics are as follows.

[benchmark data](https://docs.google.com/spreadsheets/d/1DdW0yu2x7Co91_OuFPDF0S2p76eDbYm2rFIEaSI-0Xs/edit?usp=sharing)


The current persistence core process is that the WriteBuffer data is pulled to the DataBuffer, and after certain conditions are met, it will be persisted in the Partition(IOx dimension), and the operation step is a single thread. The `persist` operation is very expensive, especially compact and write parquet_file, which is the reason why will pile up more and more data in the WriteBuffer.

```log
{"timestamp": "2022-10-31T10:20:31.188638Z", "level": "DEBUG", "fields":{"message": "Created parquet file", "num_batches":2, "num_rows": 6553700, "object_store_id": "90659f64-e53d-4879-9d3b-d0509082bf61", "partition_id": "PartitionId(85)", "write_batch_size":1024, "max_row _group_size":1048576}, "target": "parquet_file::serialize"}
{"timestamp": "2022-10-31T10:20:31.189402Z", "level": "DEBUG", "fields":{"message": "Uploading parquet to object store", "file_size": 91640446, "object_store_id": "90659f64-e53d-4879-9d3b-d0509082bf61", "partition_id": "PartitionId(85)", "total_time_to_create_parquet_ bytes": "69.42999872s"}, "target": "parquet_file::storage"}
```

Is it possible to optimize the current single-threaded processing flow into a multi-threaded concurrent process, with a concise flow as follows.

```log
   ┌───────────────┬───────────────┬───────────────┬───────────────┬───────────────┬───────────────┐              
   │               │               │               │               │               │               │              
   │    Buffer1    │    Buffer2    │    Buffer3    │   Buffer...   │     Pause     │    BufferN    │      pull    
   │     32MB      │     32MB      │     32MB      │     32MB      │               │     32MB      │  ◀───────────
   │               │               │               │               │               │               │              
   └───────┬───────┴───────┬───────┴───────┬───────┴───────────────┴───────────────┴───────────────┘              
           │               │               │                                                                      
           │               │               │                                                                      
           │               │               │         concurrent snapshot ...                                      
           │               │               │                                                                      
           │               │               │                                                                      
   ┌───────▼───────┬───────▼───────┬───────▼───────┐                                                              
   │               │               │               │                                                              
   │   Snapshot    │   Snapshot    │   Snapshot    │                                                              
   │  Persisting   │  Persisting   │  Persisting   │                                                              
   │               │               │               │                                                              
   └───────────────┴───────────────┴───────────────┘                                                              
           │               │               │                                                                      
           │               │               │                                                                      
           ▼               │               │                                                                      
max_sequence_number1       │               │                                                                      
                           ▼               │                                                                      
                 max_sequence_number2      │                                                                      
                                           ▼                                                                      
                                 max_sequence_number3                                                             
                                                                                                                  
                                                                                                                  
    ┌──────────────────────────────────────────────┐                                                              
    │                                              │                                                              
    │            Async update sort_key,            │                                                              
    │         (min)max_sequence_number,etc         │                                                              
    │                                              │                                                              
    └──────────────────────────────────────────────┘                                                              
                                                                                                                  
                                                                                                                  
                                                                                                                  
 ────────────────────────────────────────────────────────────────────────────────────────────▶  time              
```

- DataBuffer is split into multiple fixed size Buffer, so that when the maximum memory configuration is exceeded, pull data from WriteBuffer can be pause.
- A Buffer is a persist task, and the total number of persist tasks can be the number of CPU logical cores, with concurrent persistence operations to increase the consumption rate
- Synchronizing some data at regular intervals, eg: sort_key, max_sequence_number

**Comment by domodwyer** [comment]
Hey @aierui - you have been busy! 🚀 

Currently persistence is single-threaded per shard, because it consumes lots of RAM in certain situations, so running them in parallel can cause memory spikes / OOMs.

In your [document], what is `w` and `b` in the string `w100-b10`? I notice your Kafka write throughput flatlines at ~24MB/s - could writing to Kafka be the bottleneck? You can confirm this if the Kafka latency metric is high (`dml_handler_write_duration_seconds_bucket{handler="sharded_write_buffer"}`).

[document]:https://docs.google.com/spreadsheets/d/1DdW0yu2x7Co91_OuFPDF0S2p76eDbYm2rFIEaSI-0Xs/edit?usp=sharing

I do not know your system topology, but I assume you are using one Kafka topic. To parallelise the ingesters, you should create more Kafka partitions and have the ingester read from all of them - this defines the parallelism in the ingester, as it will parallelise ingest & persistence across all Kafka partitions (IOx shards).

There are also many configuration parameters for the ingester than controls how much data is read before persisting, and how often to persist - these can be tweaked to cause the ingester to persist little-but-often in the background while more writes are processed.

**Your idea is good** - parallelising persistence within a shard is a future optimisation we can make, but first we need to reduce the memory it takes to perform the persist job. We are constantly improving datafusion, and I have some prototype code changes I will add in the future that show good promise for faster persist (such as incremental compaction of snapshots before persistence, which means there will be less work for the persist operation itself - #5805). For now I am focusing on testing a Kafkaless design (as you noticed! 😄) which hopes to have even better performance and other cool things!

**Comment by aierui** [comment]
@domodwyer Thank you very much for your reply!

Let me first add something to the questions I asked. I deployed IOx system topology is:
- 1 Pod Router
- 2 Pod Ingester (5 shards per instance)
- 1 Pod Queries
- 1 Pod Compactor
- Shard Services: pg, redpanda, object store(s3)
- RedPanda: 1 topic, 10 partitions

short-word:
- w: client-worker
- b: batch-size(line protocol rows)

There are two core pain points in this issue.
1. each (IOx)Partition's data(namespace + tablename) can only be written to the same Kafka Partition 
2. an Ingester corresponds to one or more Kafka Partitions, single-threaded consumer throughput is low


But now it seems that you focusing on Kafkaless feature and round robin algorithm in `sharder` crate will solve these problems. I appreciate that you and your team are working hard to contribute advanced time-series database engine to the community. 👍

**Comment by domodwyer** [comment]
Yes writing to one table will mean writes go to one Kafka partition - throughput will be likely limited by Kafka, and ingest per Kafka partition is single threaded as you say!

It is different hardware, but I have seen a single ingester process 10,000 writes per second per shard/Kafka partition, but that is burst performance - sustained rates that need persistence will be slower.

If you want to confirm that persistence is the cause of slowness, a metric value of `ingester_paused_duration_total > 0` shows that the ingester is slowing down ingest because it is waiting for persistence to finish. If this value is not 0, you can tweak the memory config of the ingester to allow more reads before persisting.

That said, we have some cool plans that hopefully make ingest much much faster (and scalable) in general 🎉

**Comment by aierui** [comment]
> a single ingester process 10,000 writes per second per shard/Kafka partition

Can you share some more data here? For example, metrics/s, record rows/s, fields/row, tags/row

> we have some cool plans that hopefully make ingest much much faster

I am very much looking forward to your new features.

**Comment by domodwyer** [comment]
So I have recently worked to improve this in `ingester2`; I'd like to share an update with you!

As you saw in #6316, I have changed partitions so they can have multiple buffers in a persist queue - this lets us have multiple outstanding persist jobs per partition. I will soon follow this up with a new consistent hashing persist system that serialises buffer persistence within a partition, but parallelises persistence across partitions (so all persist ops for partition 1 are serialised, but they can run in parallel with partition 2/3/4/etc).

Because the partition sort key MAY be updated when a partition is persisted, persist operations for a single partition cannot be trivially parallelised - the sort key updates are  not commutative, so they have to be serialised in order to produce the correct end result. There is an opportunity here though - if a partition does not need to update the sort key, it can be parallelised!

I think the new persist system will help with performance overall, but I will leave this ticket open to investigate parallelising partition persistence.
